specifyweb/frontend/js_src/lib/components/Merging/InvalidMergeRecords.tsx

```diff
@@ -13,32 +13,40 @@ import { commonText } from '../../localization/common';
 import { mergingText } from '../../localization/merging';
 import { Tables } from '../DataModel/types';
 import { TableIcon } from '../Molecules/TableIcon';
+import { className } from '../Atoms/className';
 
 function InvalidMergeRecords({
   resources,
   tableName,
-  body = '',
-  specificText,
+  specificText = '',
 }: {
   readonly tableName: keyof Tables;
   readonly resources: RA<SerializedResource<AnySchema>>;
-  readonly body?: string;
   readonly specificText?: string;
 }): JSX.Element {
   return (
     <div className="flex flex-1 flex-col gap-4">
-      <H2>{body}</H2>
+      <H3 className={className.headerPrimary}>
+        {mergingText.recordNotBeMergedReason()}
+      </H3>
       <p>{specificText}</p>
       <Ul className="gap-2">
-        {resources.map((resource, index) => (
-          <li
-            key={(resource.id as number) ?? index}
-            className="flex min-h-[theme(spacing.8)] flex-1 items-center gap-2"
-          >
-            <TableIcon label={true} name={tableName} />
-            <FormattedResource resource={deserializeResource(resource)} />
-          </li>
-        ))}
+        {resources.map((resource, index) => {
+          return (
+            <li
+              key={(resource.id as number) ?? index}
+              className="flex min-h-[theme(spacing.8)] flex-1 items-center gap-2"
+            >
+              <TableIcon label={true} name={tableName} />
+              <FormattedResource
+                resource={React.useMemo(
```

Comment:
what did we learn about hooks?
hooks can't be used in here! 🤣
move this useMemo to the top level
and enable ESLint so it can shout about this, instead of me shouting about this  🤣

Reply:
I thought it was an anti-pattern but couldn't find a reason why this isn't allowed in this specific case. Maybe because I could have different number of hook calls between renders?

Reply:
It's not an anti-pattern. It's simply not allowed.

There are only 3 rules of hooks and it pays to have them well memorized:
https://www.w3schools.com/react/react_hooks.asp#:~:text=Hook%20Rules,Hooks%20cannot%20be%20conditional

> Hooks can only be called inside React function components.
> **Hooks can only be called at the top level of a component.**
> Hooks cannot be conditional

Pretty sure @CarolineDenis tried to violate each of these at one point 🤗

Reply:
:joy::see_no_evil:

Reply:
Solutions:
- First thing that came to mind, but not the best: Move useMemo outside of the map. and useMemo will have a separate map inside of it, creating an array, which would be accessed by the .map that is in the return statement
- Best performance and the cleanest: extract the contents of the .map() into a separate component. Inside of that component, useMemo hook can be called (as it would be at the top level)
- A bit weird, and not best performance (because on change to any resource, all resources will be re-computed): `const li = React.useMemo(()=> /* copy the entire .map() as you have it right now, except remove useMemo() that you have inside of .map() right now */ ,[...]);`

Reply:
Resolved. Another possible solution might be to use logic similar to useResource uses https://github.com/specify/specify7/blob/a4188f0e9b0b43f8188302698fec2fabc0678e17/specifyweb/frontend/js_src/lib/components/Merging/index.tsx#L404